Block notification service in sandbox

pvollan opened the pull request:
#### 01a6cf5f9913b9794742ea236abf2c4369a6d794
<pre>
Block notification service in sandbox
<a href="https://bugs.webkit.org/show_bug.cgi?id=270805">https://bugs.webkit.org/show_bug.cgi?id=270805</a>
<a href="https://rdar.apple.com/124395477">rdar://124395477</a>

Reviewed by Sihui Liu and Chris Dumez.

Block notification service in the WebContent process sandbox.

* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.WebContent.sb.in:
* Source/WebKit/Scripts/process-entitlements.sh:
* Source/WebKit/Shared/Sandbox/common.sb:
* Source/WebKit/WebProcess/com.apple.WebProcess.sb.in:

Canonical link: <a href="https://commits.webkit.org/277077@main">https://commits.webkit.org/277077@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2914df430caee0a1cdaeb45fc7fc5e5a8795032a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/46121 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/25252 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/48706 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/48792 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/42161 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/29575 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/22694 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/37706 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/46699 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/22339 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/39783 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/18901 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/45986 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/19737 "Passed tests") | | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/4165 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/42437 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/41230 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/50586 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/21117 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/17631 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/44874 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/22417 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/43776 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/22776 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6516 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/22111 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WTF/wtf/PlatformEnableCocoa.h:
```diff
@@ -1036,3 +1036,9 @@
+
+#if !defined(ENABLE_NOTIFY_BLOCKING) \
+    && (PLATFORM(MAC) && __MAC_OS_X_VERSION_MIN_REQUIRED >= 150000) \
+    || (PLATFORM(IOS) && __IPHONE_OS_VERSION_MAX_ALLOWED > 170400)
```
What about WatchOS, TVOS?

pvollan replied:
That is a very good point. Will add!

Thanks for reviewing!

pvollan replied:
Actually, we may want to enable these at a later point here.

bnham commented on Source/WTF/wtf/PlatformEnableCocoa.h:
```diff
@@ -1036,3 +1036,7 @@
 #define ENABLE_QUICKLOOK_SANDBOX_RESTRICTIONS 1
 #endif
+
+#if !defined(ENABLE_NOTIFY_BLOCKING) && PLATFORM(IOS) && __IPHONE_OS_VERSION_MAX_ALLOWED > 170400
```
Is the version number correct here?

pvollan replied:
Yes, looking closer, this is intentional.

Thanks for reviewing!

szewai commented on Source/WTF/wtf/PlatformEnableCocoa.h:
```diff
@@ -1036,3 +1036,7 @@
 #define ENABLE_QUICKLOOK_SANDBOX_RESTRICTIONS 1
 #endif
+
+#if !defined(ENABLE_NOTIFY_BLOCKING) && PLATFORM(IOS) && __IPHONE_OS_VERSION_MIN_REQUIRED >= 180000
```
Do we want to check against target build instead of SDK?

szewai reviewed (approved):
r=me

szewai commented on Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.WebContent.sb.in:
```diff
@@ -503,39 +502,37 @@
-
-#if ENABLE(NOTIFY_BLOCKING)
-(with-filter (state-flag "EnableExperimentalSandbox")
+(define (allow-notifyd-messages)
```
It seems `(allow ipc-posix-shm-read* (ipc-posix-name "apple.shm.notification_center"))` could be included in this

pvollan replied:
Ah, that is a good point, will fix!

Thanks for reviewing!

szewai commented on Source/WebKit/Scripts/process-entitlements.sh:
```diff
@@ -186,6 +186,20 @@ function webcontent_sandbox_entitlements()
+function notify_entitlements()
+{
+    plistbuddy Add :com.apple.developer.web-browser-engine.restrict.notifyd bool YES
+    plistbuddy Add :com.apple.developer.web-browser-engine.webcontent.restrict.notifyd bool YES
```
Do we need both entitlements? What is the difference?
Asking this as it seems we only check for the webcontent one in `notify-blocking`.

szewai commented on Source/WebKit/WebProcess/com.apple.WebProcess.sb.in:
```diff
@@ -1208,33 +1205,28 @@
+(allow ipc-posix-shm-read* (ipc-posix-name "apple.shm.notification_center"))
+
+#if ENABLE(NOTIFY_BLOCKING)
+(when (equal? (param "ENABLE_SANDBOX_MESSAGE_FILTER") "YES")
```
Would there be a behavior difference for changing `if` to `when`?

pvollan replied:
That's a good point. In this case there would be a behavior difference, since `if` has an `else` statement. 

```
(if (condition)
    (statement1)
    ;; else
    (statement2))
```

szewai commented on Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.WebContent.sb.in:
```diff
@@ -503,39 +502,37 @@
+
+#if ENABLE(NOTIFY_BLOCKING)
+(with-filter (notify-blocking)
+    (deny mach-lookup (global-name "com.apple.system.notification_center")))
```
Based on code, it seems if we don't explicitly `allow mach-lookup (global-name "com.apple.system.notification_center"))`, the default behavior is deny, right?
Then this should not be necessary?

pvollan replied:
Good point, will fix!

szewai commented on Source/WebKit/WebProcess/com.apple.WebProcess.sb.in:
```diff
@@ -1208,33 +1205,28 @@
-                (deny mach-message-send (with telemetry-backtrace) (notifyd-message-numbers))))))
+;; FIXME should be removed when <rdar://problem/9347205> + related radar in Safari is fixed
+(if (equal? (param "ENABLE_SANDBOX_MESSAGE_FILTER") "NO")
+    (allow mach-lookup (global-name "com.apple.system.notification_center"))
```
This rule is in `allow-notifyd`, so this is unnecessary?

pvollan replied:
In this case I don't believe we can use `allow-notifyd-messages`, since the message filter syntax is not allowed when `ENABLE_SANDBOX_MESSAGE_FILTER ` is `NO`.

szewai commented on Source/WebKit/WebProcess/com.apple.WebProcess.sb.in:
```diff
@@ -1208,33 +1205,27 @@
+        (deny mach-lookup (global-name "com.apple.system.notification_center"))
+        (deny ipc-posix-shm-read* (ipc-posix-name "apple.shm.notification_center")))
+    (with-filter (require-not (notify-blocking))
+        (allow-notifyd-messages)))
```
This should be `allow-notifyd`since it is renamed

pvollan replied:
Ah, good catch, will fix!

szewai commented on Source/WebKit/WebProcess/com.apple.WebProcess.sb.in:
```diff
@@ -1208,33 +1205,27 @@
+(if (equal? (param "ENABLE_SANDBOX_MESSAGE_FILTER") "NO")
+    (allow mach-lookup (global-name "com.apple.system.notification_center"))
+;; else
+    (allow-notifyd))
```
Can we merge the next rules here, something like:
```
if ENABLE(NOTIFY_BLOCKING)
    (with-filter (require-not (notify-blocking))
         (allow-notifyd)))
#else
    (allow-notifyd))
#endif
```
Then we don't need the deny rules?

pvollan replied:
Good point, will fix!

pvollan committed:
Block notification service in sandbox
https://bugs.webkit.org/show_bug.cgi?id=270805
rdar://124395477

Reviewed by Sihui Liu and Chris Dumez.

Block notification service in the WebContent process sandbox.

* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.WebContent.sb.in:
* Source/WebKit/Scripts/process-entitlements.sh:
* Source/WebKit/Shared/Sandbox/common.sb:
* Source/WebKit/WebProcess/com.apple.WebProcess.sb.in:

Canonical link: https://commits.webkit.org/277077@main
